### DISCUSSION TITLE
Is zfs_autobackup right for me?

### DISCUSSION THREAD

**Discussion by danboid** [psy0rz/zfs_autobackup#226] [Q&A] [answered]
I have read the documentation through a couple of times now but it leaves me uncertain about how zfs_autobackup prefers / handles the creation of snapshots?

It sounds like it can (must?) be used without using another program such as zfs-auto-snapshot or sanoid to create automated snapshots but what if the user wants to use z-a-s (read: I am already using and trust z-a-s) on the source machine instead of zfs_autobackup creating its own snapshots first every time it is run? I can't see an option to only transfer existing snapshots?

I want to use zfs-auto-snaphot on my source machine to create the snapshots but in the zfs_autobackup docs it seems to imply that zfs_autobackup will create a new snapshot (in every labelled dataset?) every time is run in a non-test mode? Can this be disabled so that it never creates new snapshots on the source?

I don't want zfs_autobackup to create any snapshots at all on the source machine, I just want it to transfer all of the latest snapshots from all of the datasets on the source machine onto the destination machine and additionally I want it to thin out all snapshots in all datasets on the destination machine only so that only keeps 90 days / 12 weeks worth of snapshots in all of the datasets. z-a-s handles the destroying of snapshots on the source side so no need for zfs_autobackup to do any thinning on the source.

I am hoping the thinning works reliably on the destinstion side. sanoid's auto-pruning didn't work at all for me which is why I'm looking into this as an alternative.

Another question is regarding `--strip-path=1`. The docs for this option  say:

    Select the whole source-pool. In this case: zfs set autobackup:offsite1=true rpool
    Use --strip-path=1
    Specify target-pool as target-path. In this case: data
    Use the --force option the first time to overwrite the existing target pool. (New in v3.1.2)

I want to recursively replicate all of the datasets in /pool/home on the source into /pool2/home on the destination so it sounds like I need to use `--strip-path=1`. 

Why then should I need to label/select the whole pool on the source if its only `/pool/home` that I want to recursively replicate, not the whole pool? Do I really need to go to the trouble of setting all of the other datasets in the source pool to false just so I can use `--strip-path=1` on one dataset, the one I actually want to replicate without its path.

Finally, I presume that zfs_autobackup can work, on the destination side at least, as non-root users if we delegate the required permissions to the correct user? What are the minimum set of zfs permissions required for it to work on either side?

Thanks for your help!

**Comment by psy0rz** [comment, marked as answer]
thats a lot of questions, and some of them you have to figure out and --test yourself. :)

you can use --no-snapshot to not create snapshots

you can --keep-source=10000 to not delete snapshots from the source


also the snapshots should have to correct format, or use --snapshot-format. otherwise use --other-snapshot and prune them yourself on the target.


this all sounds like too much of a hassle to be honest..just let zfs-autobackup create its own snapshots to make your life easier. You can still use zfs-auto-snapshot, zfs-autobackup will just ignore those.

**Reply by danboid** [reply]
Thanks for the quick reply @psy0rz !

Can I honestly use zfs-autobackup to reliably create automated snapshots for a pool with over 800 datasets (home dirs) with 4 frequent (every 15 minutes), 6 hourly, 12 weekly and 3 monthly snapshots for all 800+ datasets? Thats what z-a-s manages for me and its reliable. I'm pretty sure I'll have to configure --snapshot-format to make zfs-backup use the z-a-s format because I doubt zfs-autobackup will be able to do that and also do nightly syncs, thinning on the destination, will it? Has anyone tested zfs_autobackup at this scale before?

**Reply by danboid** [reply]
Can I issue multiple values to --snapshot-format because thats what I'd need to do to use it with z-a-s unless I revert to only creating daily snapshots with it.

**Reply by danboid** [reply]
Actually, it might make more sense for me to modify the zas scripts so that they use the correct format used by zfs_autobackup as I dont think it zfs_autobackup supports setting multiple snapshot formats does it? I'd need one for frequent, one for daily, one or weekly and one for monthly. It sounds like zfs_autobackup does do everything I need it to otherwise.

If you know that zfs_autobackup can reliably create as many snapshots at the scale I'm dealing with better than z-a-s then I'll drop z-a-s but it doesn't seem that it was designed to handle this kind of scale?

Please excuse my barrage of questions. This looks like it would be my best choice of ZFS snapshot replication tool if I can get it to co-exist with / manage z-a-s snapshots. I would need zfs_autobackup to trim all of the old snapshots on the destination.

**Reply by psy0rz** [reply]
ahh ok i see. 

you can indeed only use one snapshot-format.

also, the amount is no problem, see: https://github.com/psy0rz/zfs_autobackup/wiki/Performance#some-statistics

However, the select-stage, during which zfs-autobackup checks which datasets have the autobackup:... property might be a bit IO intensive. It does a `zfs get -t volume,filesystem -o name,value,source -H autobackup:backup`

If you follow the performance tips from the wiki, i think that the rest of the steps shouldn't be a problem.

**Reply by psy0rz** [reply]
the performance tips are here: https://github.com/psy0rz/zfs_autobackup/wiki/Performance

The "speeding up SSH" and "less zfs commands" are a must, the rest of the tips depends on your situation.

**Reply by danboid** [reply]
It would be great if you could add a z-a-s compatible thinning feature for people who want to / have been using that for snapshot creation. I might look into adding such a feature if I end up using this and you think that would be doable?

I would prefer to be able to use zfs_autobackup to transfer all of my z-a-s created snapshots but I may forgo that idea and let zfs_autobackup create its own snapshot(s) once a day. I could then let zfs_autobackup manage its own snapshots on both ends and just have the z-a-s snaps existing independently on the source machine I suppose? It isn't such a big deal if we only have daily snapshots on the backup server and lose the houly and frequent snapshots I suppose. I would prefer to retain them too ideally.

It wasn't my plan but I like the simplicity of using zfs_autobackup. I like that it uses python instead of perl or go or whatever.

**Reply by psy0rz** [reply]
i think that 3rd-party compatible thinning feature is a bit too domain specific. i try to keep it general and simple, which can be difficult enough already. :)

also: all code needs to be regression tested, which makes adding such features ever more challenging.

**Reply by psy0rz** [reply]
well, maybe --snapshot-format can be changed/extended so that it supports a list of formats. (and uses the first format when creating snapshots)

**Reply by psy0rz** [reply]
i'm all for that, and then the wiki can contain a list of  "snapshot-formats compatible with tool X" or something.

**Reply by danboid** [reply]
Do you know a reliable ZFS snapshot replication tool that might be a better fit for me, one that would replicate and thin all of my z-a-s snapshots? sanoid's auto-prune code doesn't scale to 800 datatsets and Ive read of others having the same issue with much fewer datasets involved.

I may well use zfs_autobackup in the way I described. Retaining more snspshots would be nice but thats not a showstopper and not totally essential.

**Reply by psy0rz** [reply]
nope i dont know such a tool.

i would say: just try zfs-autobackup with the performance tips, and let me know how well it works. :P

**Reply by danboid** [reply]
I'm going to investigate zrepl. It seems to have more pruning options than zfs_autobackup but it may fail me elsewhere?

https://zrepl.github.io/configuration/prune.html#prune

**Reply by psy0rz** [reply]
Its a very advanced and powerfull tool, but i didnt look into the snapshot-formatting support yet.

However keep in mind that snapshots are not atomic in zrepl.

**Reply by danboid** [reply]
I don't want to create any snapshots with zrepl so as long as it can replicate atomically created snapshots then I should be OK.

zrepl looks trickier to configure than zfs_autobackup and syncoid, znapzend etc but at least it does support manual snapshot creation and so it must support z-a-s created snapshots, I presume.

https://zrepl.github.io/configuration/snapshotting.html#manual-snapshotting

**Reply by psy0rz** [reply]
Yes, its powerfull but has a high learning curve and more complex setup process.

**Reply by danboid** [reply]
Yes, I've been reading the docs more and I think it would take some time to get configured.

Two other advantages that autobackup has are the it has a --test mode and it seems zrepl requires that you install it on both the receiver and the sender. I don't think that getting to retain a few more snapshots is worth all that extra hassle.

**Reply by danboid** [reply]
I'm giving autobackup a proper test now. Its only a few datasets in but I've already seen these errors a few times:

```
[Target] STDERR > cannot unmount '/home': pool or dataset is busy
! [Source] Command "ssh zfsbackup@X.X.X.X 'zfs send -L -e -c -v -P -p astarray/home@asteria-homedirs-20231012231530'" returned exit code 141 (valid codes: [0])
! [Target] Command "zfs recv -u -o canmount=noauto -v -F -s letome/home" returned exit code 1 (valid codes: [0])
! [Source] astarray/home: FAILED: Last command returned error
```

It seems to be working fine apart from that.

If I use the `--strip-path=1` option and I just use the destinations pool name as the target, will it skip replicating any datasets that I run `zfs set autobackup:asteria-homedirs=false` on? If so (I hope so), this should be added to the --strip-path section of the README.

**Reply by danboid** [reply]
autobackups --keep-source settings only thin the snapshots that it creates on the source right or will it also thin the z-a-s snaps on the source?

Yesterday I went nuclear on the snapshots to start over with snapshotting so I don't have any snapshots older than a day right now to tell if its thinning > 3 month old z-a-s snaps. I don't want it to but I do want it to thin the autobackup snaps it creates on the source and target.

EDIT

Come to think of it, it doesn't really matter if autobackup is trimming all snapshots o the source as long as that doesn't cause probs for z-a-s but I can't see it. I want to keep the same period of snapshots (3 months) for both z-a-s snaps and autobackup snaps.

**Reply by danboid** [reply]
It seems that my first proper test of autobackup has worked except it didn't mount any of the datasets on the target after replicating them.

Can I configure autobackup to auto mount all new datasets that it replicates on the target? I suppose I could add `zfs mount -a` as a post backup command.

**Reply by psy0rz** [reply]
Yea, this is a known issue that is fixed in master. See #112

**Reply by psy0rz** [reply]
Zfs-autobackup only thins it own snapshots. (Those that match the snapshot-format)

Even if you use --other-snapshots

**Reply by danboid** [reply]
OK. I've tried running `zfs mount -a` and rebooting it but non of my freshly imported datasets are getting mounted.

I know how to manually fix this but obviously I'd like to avoid doing that if we can get zfs-autobackup to do it for me.

If I upgrade to master, what would I need to do to fix the mount points? Would I have to destroy and re-send all of the datasets to get it to work properly?

**Reply by danboid** [reply]
If I use a script to fix all the mount points (running `zfs set mountpoint=/home/user astarray/home/user` in a for loop), will changing the mountpoints cause any problems for future runs of autobackup? Obviously I'd prefer to fix it the way zfsautobackup prefers because I'd like new datasets to get automounted.

**Reply by psy0rz** [reply]
normally the original mountpoint should be preserved, so a zfs mount -a should work. what is their mountpount before you explicitly set it ?

**Reply by danboid** [reply]
If I run `zfs list -o name,mountpoint` all of the mount points are present and correct but nothing is mounted. I've already running `zfs mount -a` but that didn't mount any of them.

I think this might be because the canmount property is set to `noauto` of all of the child datasets. Do you think setting `canmount` to `on` for all of the datasets would fix it?

How would I get autobackup to set `canmount` to `on` by default if this is my problem? All of the source datasets have `canmount` set to `on`.

**Reply by psy0rz** [reply]
yes you should set canmount to on.

did you use the --clear-mountpoint parameter? also please read https://github.com/psy0rz/zfs_autobackup/wiki/Mounting

**Reply by psy0rz** [reply]
normally zfs-autobackup doesnt modify any properties, so the canmount should be the same as on the source.

**Reply by danboid** [reply]
Looks like a bug then? The canmount property was set to noauto on of all the target datasets after replication. The command I used was:

```zfs-autobackup -v --clear-mountpoint --strip-path=1 --keep-source=1d90d,1w12w --keep-target=1d90d,1w12w --zfs-compressed --ssh-source zfsbackup@X.X.X.X asteria-homedirs letome```

Ran as root on the target with autobackup 3.2. We're running Ubuntu 22.04 on the source and destination.

**Reply by danboid** [reply]
`canmount` was set to `on` on both the source parent dataset (/home) and the target dataset.

**Reply by psy0rz** [reply]
You used --clear-mountpoint ..

**Reply by danboid** [reply]
Oh OK, the docs recommend using that but it seems in my case I shouldn't be using it at all?

**Reply by psy0rz** [reply]
Depends on the usecase

**Reply by danboid** [reply]
Done my second test now it all looks good except...

When I enable verbose mode (-v), it first lists the selected datasets then in the 

#### Snapshotting

Section it says `No changes since` for EVERY dataset which is bollox because when it gets to

#### Synchronising

We can see that several datasets did have changes because several receive incremental updates.

It looks like the snapshotting section isn't working or at least isn't reporting the correct status of modified datasets?

**Reply by danboid** [reply]
My initial 20 questions feet to the fire bit is over with now BTW, I hope you enjoyed it?

Thanks for being prompt with your replies, I will likely tweak your docs soon in return. They are already good actually which is another reason to choose the autobackup.

What do you call it anyway? zab?

**Reply by psy0rz** [reply]
It skips snapshot creation if there are no changes, but it will still sync snapshots there are already created.

**Reply by psy0rz** [reply]
I usually call it zfs-autobackup. :)

The questions are fine, whenever possible i try to improve either the program or documentation, to prevent such questions in the future.

**Reply by danboid** [reply]
I have added the explantion of --no-snapshot to the manual.

**Comment by danboid** [comment]
If a user is doing using zab for pull backups and didn't want to let zfs-autobackup create any snapshots on the source machine, does zfs-autobackup still work fine for replicating existing snapshots to the target? Obviously zfs-autobackup's source snapshot trimming command becomes useless in such cases.

**Reply by danboid** [reply]
I'll be integrating your responses into the documentation shortly.

**Reply by danboid** [reply]
I've finished switching us from syncoid to zfs-autobackup now for our GROUNDBREAKING ZFS LTSP server :D

**Comment by viharm** [comment]
Great discussion. Thanks @danboid & @psy0rz .
